pkg/proxy/dns.go

```diff
@@ -48,6 +48,8 @@ func (dr *dnsRedirect) setRules(wg *completion.WaitGroup, newRules policy.L7Data
 	}
 	dr.currentRules = copyRules(dr.redirect.rules)
 
+	log.Infof("Updating the DNS rules for endpoint %d", dr.redirect.endpointID)
+	dr.redirect.localEndpoint.SyncEndpointHeaderFile()
```

Comment:
`SyncEndpointHeaderFile` is already called within the endpoint pkg (`pkg/endpoint`) and during policy recalculation. Why is this necessary here especially expanding the scope of the proxy to reach into the management of endpoints?

Reply:
Hey sorry for the delay in response. So the intention here is to update the DNS rules to the `ep_config.json` when we apply the policy.
Afaik, it does not write the dns rules to the file when the policy is applied. So I was calling this function which is triggered when we want to write the dns rules to the file. 
Current implementation, I see that the data is written to the fs after the policy applied, but dns rules are not. Is there another way we can incorporate getting the DNS rules to be written to the fs ?

Reply:
> Current implementation, I see that the data is written to the fs after the policy applied, but dns rules are not. Is there another way we can incorporate getting the DNS rules to be written to the fs ?

I'm not sure I understand the problem you are describing. The DNS rules are in fact written to the fs as they are fetched from the same function: https://github.com/cilium/cilium/blob/325b5df02179d3ae9c62d074aadd07636346f1fb/pkg/endpoint/endpoint.go#L2438

Reply:
Also another note to keep in mind here, see the comment regarding locks and deadlocks on the line above of the linked code: this PR might further complicate the locking pattern by now having the proxy reach into the headerfile.

Reply:
I agree; this is probably not necessary here. Rather, we should be syncing the endpoint header file explicitly on regeneration, which IIRC is not done now.